src/Coordination/Changelog.cpp

```diff
@@ -540,23 +546,62 @@ ChangelogRecord Changelog::buildRecord(uint64_t index, const LogEntryPtr & log_e
     return record;
 }
 
+void Changelog::writeThread()
+{
+    // we don't protect current_writer because rotate at the same time can be called from compact only
+    // when the node is applying snapshot from leader
+    // in that case, no entry write should happen, i.e. this thread will be inactive
```

Comment:
To me it's hard to validate that this invariant is true :(
Also, I cannot validate that `current_writer` exists if there is an `AppendLog` operation in `write_operations`.

Maybe it can be easier to understand if we create a separate class for write operations queue, which owns `current_writer`. (Or something similar).

Reply:
Maybe it's easier to just protect it with mutex?
I've just noticed that `write_at` will also use `current_writer` which can be called in parallel while `writeThread` is doing `rotate`.